firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java

```diff
@@ -469,12 +469,35 @@ public void notifyLocalViewChanges(List<LocalViewChanges> viewChanges) {
         "notifyLocalViewChanges",
         () -> {
           for (LocalViewChanges viewChange : viewChanges) {
-            localViewReferences.addReferences(viewChange.getAdded(), viewChange.getTargetId());
+            int targetId = viewChange.getTargetId();
+
+            localViewReferences.addReferences(viewChange.getAdded(), targetId);
             ImmutableSortedSet<DocumentKey> removed = viewChange.getRemoved();
             for (DocumentKey key : removed) {
               persistence.getReferenceDelegate().removeReference(key);
             }
-            localViewReferences.removeReferences(removed, viewChange.getTargetId());
+            localViewReferences.removeReferences(removed, targetId);
+
+            if (!viewChange.hasUnresolvedLimboDocuments()) {
+              QueryData queryData = targetIds.get(targetId);
+              hardAssert(
+                  queryData != null,
+                  "Can't set limbo-free snapshot version for unknown target: %s",
+                  targetId);
+
+              // Advance the last limbo free snapshot version
+              SnapshotVersion lastLimboFreeSnapshotVersion = queryData.getSnapshotVersion();
+              QueryData updatedQueryData =
```

Comment:
This seems like it would be more readable if we added copy-setters for the things we update, something like `queryData.withLastLimboFreeSnapshotVersion(lastLimboFreeSnapshotVersion)`.

What do you think?

Reply:
I will send you a follow-up that updates the code here and elsewhere.